File: src/iot/utility/log_utils.py
import datetime
import sys
from zoneinfo import ZoneInfo

# TODO: Fix logging (program_name)
# TODO: Find and remove libraries that are unnecessary

PROGRAM_NAME = None

def set_program_name(program_name):
    global PROGRAM_NAME
    PROGRAM_NAME = program_name

def print_log(format_str, *args):
    """
    Logs messages with a timestamp and program name.

    Args:
        program_name (str): The name of the program or module.
        format_str (str): The format string for the message.
        *args: Variable length argument list to format the string.

    Returns:
        None
    """
    current_time = datetime.datetime.now(ZoneInfo('US/Eastern'))
    timestamp = current_time.strftime('%Y-%m-%d %H:%M:%S')

    if format_str.startswith('\n'):
        # add newline before the timestamp
        print(f"\n{timestamp}::{PROGRAM_NAME}::", end='')
        format_str = format_str[1:]  # remove the first character
    else:
        print(f"{timestamp}::{PROGRAM_NAME}::", end='')

    print(format_str.format(*args))


def perror_log(format_str, *args):
    """
    Logs error messages with a timestamp and program name to stderr.

    Args:
        program_name (str): The name of the program or module.
        format_str (str): The format string for the error message.
        *args: Variable length argument list to format the string.

    Returns:
        None
    """
    current_time = datetime.datetime.now(ZoneInfo('US/Eastern'))
    timestamp = current_time.strftime('%Y-%m-%d %H:%M:%S')
    print(f"{timestamp}::{PROGRAM_NAME}::{format_str.format(*args)}", file=sys.stderr)

File: src/iot/utility/log_utils.c
#include "log_utils.h"
#include "utils.h"
#include <time.h>
#include <stdlib.h>
#include <stdio.h>

// TODO: Fix logging (program_name)
// TODO: Find and remove libraries that are unnecessary

/**
 * @brief Prints a formatted log message with a timestamp and program name.
 * @param program_name The name of the program or module.
 * @param format The format string for the log message.
 * @param ... Additional arguments for the format string.
 */
void print_log(const char *program_name, const char *format, ...) {
    va_list argptr;
    va_start(argptr, format);

    char timestamp[32];
    get_current_time_in_est(timestamp, "%Y-%m-%d %H:%M:%S");

    if (format[0] == '\n') {
        printf("\n%s::%s::", timestamp, program_name);
        format++;
    } else {
        printf("%s::%s::", timestamp, program_name);
    }

    vprintf(format, argptr);

    va_end(argptr);
}

/**
 * @brief Prints an error log message with a timestamp, program name, and additional system error details.
 * @param program_name The name of the program or module.
 * @param format The format string for the log message.
 * @param ... Additional arguments for the format string.
 */
void perror_log(const char *program_name, const char *format, ...) {
    va_list argptr;
    va_start(argptr, format);

    char timestamp[32];
    get_current_time_in_est(timestamp, "%Y-%m-%d %H:%M:%S");

    fprintf(stderr, "%s::%s::", timestamp, program_name);
    vfprintf(stderr, format, argptr);

    perror("");

    va_end(argptr);
}

/**
 * @brief Prints a hexadecimal representation of the given data.
 * @param program_name The name of the program or module.
 * @param pbtData Pointer to the data.
 * @param szBytes Size of the data in bytes.
 */
void print_hex(const char *program_name, const uint8_t *pbtData, const size_t szBytes) {
    if (!pbtData || !program_name) {
        print_log(program_name, "Invalid input to print_hex");
        return;
    }

    // Ensure szBytes isn't too large to prevent potential stack overflow.
    if (szBytes > 1024) {  // Arbitrary limit, adjust as needed
        print_log(program_name, "Data too large to process in print_hex");
        return;
    }

    char message[szBytes * 3 + 1];  // 2 characters for hex + 1 for space
    char *current = message;

    for (size_t szPos = 0; szPos < szBytes; szPos++) {
        current += sprintf(current, "%02x ", pbtData[szPos]);
    }

    // Replace the last space with a newline
    if (szBytes > 0) {
        message[szBytes * 3 - 1] = '\n';
    }

    message[szBytes * 3] = '\0';  // Null-terminate the string
    print_log(program_name, message);
}